pkg/podmounter/mppod/headroom.go

```diff
@@ -0,0 +1,170 @@
+package mppod
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"slices"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+)
+
+// Labels populated on spawned Headroom Pods.
+const (
+	LabelHeadroomForPod    = "experimental.s3.csi.aws.com/headroom-for-pod"
+	LabelHeadroomForVolume = "experimental.s3.csi.aws.com/headroom-for-volume"
+)
+
+// Labels populated on Workload Pods requesting Headroom Pods.
+const (
+	LabelHeadroomForWorkload = "experimental.s3.csi.aws.com/headroom-for-workload"
+)
+
+// A scheduling gate can be used on Workload Pods using a volume backed by the CSI Driver to signal the CSI Driver
+// to reserve headroom for the Mountpoint Pod to serve volumes to workload.
+//
+// If this scheduling gate is used on a Workload Pod, the CSI Driver:
+//  1. Labels the Workload Pod to use inter-pod affinity rules in the Headroom Pods
+//  2. Creates Headroom Pods using a pause container with inter-pod affinity rule to the Workload Pod
+//  3. Ungates the scheduling gate from the Workload Pod to let it scheduled - alongside the Headroom Pods if possible
+//  4. Schedules Mountpoint Pod if necessary (i.e., the CSI Driver cannot share an existing Mountpoint Pod) into the same node as the Workload and Headroom Pods using a preempting priority class
+//  5. Mountpoint Pod most likely preempts the Headroom Pods if there is no space in the node - as the Headroom Pods uses a negative priority -, or just gets scheduled if there is enough space for all pods
+//  6. Deletes the Headroom Pods as soon as the Workload Pod is running or terminated - as Mountpoint Pods are already scheduled or no longer needed
+const SchedulingGateReserveHeadroomForMountpointPod = "experimental.s3.csi.aws.com/reserve-headroom-for-mppod"
```

Comment:
Not sure that we should use prefix `experimental` in labels and scheduling gate names.

E.g. if this feature would become stable in the future we might need to maintain 2 types of labels and support some kind of migration code to ensure we are not breaking anything during stable release.

I think even K8s never uses `experimental` in their labels for experimental features.

Reply:
I can only see one experimental label in https://kubernetes.io/docs/reference/labels-annotations-taints/, but there are lots of alpha and beta labels/annotations.

But your point makes sense, I don't expect us to change label/scheduling-gate names between versions, and we can drop experimental there as we already have experimental feature-flag at Helm-level.